### DISCUSSION TITLE
Running the RoboticsAcademy exercise on a local machine

### DISCUSSION THREAD

**Discussion by BkPankaj** [JdeRobot/VisualCircuit#284] [Q&A] [answered]
Hello everyone,

I am seeking assistance with solving Robotics Academy exercises using VisualCircuits. I have watched some videos demonstrating VisualCircuits, where launch and world files are launched, such as in one [video](https://youtu.be/J1JhnNOKe1o?si=Ke_lHrM4FtXfDygu) where the visual_lander.launch file is used. However, in the recent version of RoboticsAcademy, the file organization has changed, and in most folders, there are no world and launch files available for testing on a local machine. In exercises like Amazon Warehouse, launch files and world files are provided. I am attempting to run the autoparking exercise on my local machine but I am uncertain about the steps required for local execution. Could anyone please guide me on how to run the exercises on my local machine?

**Comment by toshan-luktuke** [comment]
Hi @BkPankaj which exercise are you trying to access exactly?

**Reply by BkPankaj** [reply]
Hello @toshan-luktuke ,
I am trying to replicate the [visual lander](https://jderobot.github.io/RoboticsAcademy/exercises/Drones/visual_lander) as demonstrated in the above video, while my primary focus on the [autoparking](https://jderobot.github.io/RoboticsAcademy/exercises/AutonomousCars/autoparking/) exercise on ROS2 Humble.

**Reply by toshan-luktuke** [reply, marked as answer]
Drones with Robotics Academy actually require a lot of auxiliary dependencies. I wrote a [blog](https://theroboticsclub.github.io/gsoc2022-Toshan_Luktuke/gsoc/installing-jde-drone/) describing how I got it running locally. You can refer to that and see if it helps. Aside from that, most of the exercises are limited only by the dependencies, i.e. if all dependencies are there, they should run normally using the main.py files (since all they do is run the launch files)

Alternatively, I would suggest running the Docker container for RoboticsAcademy, copying the VisualCircuit Package (`docker cp`) that is made and then running it normally. It will work as expected as long as the topics are up and running for the worlds.
So the process would be something like:
1. Start RoboticsAcademy Image
2. Start the Visual Lander exercise via Web-Interface or command line
3. Copy Files from the VisualCircuit package to the running container
4. Run the Visual Circuit Program using `python3 main.py`

**Reply by BkPankaj** [reply]
Thank you for reply and sharing the blog. I will try the Docker method, and I will let you know about the updates.

**Reply by BkPankaj** [reply]
Thank you, @toshan-luktuke, for your support. I followed your steps, and the Docker method worked perfectly.

**Comment by BkPankaj** [comment]
Greetings everyone,

I successfully implemented the autoparking exercise of RoboticsAcademy using VisualCircuits. I followed the steps outlined by @toshan-luktuke in [this comment](https://github.com/JdeRobot/VisualCircuit/discussions/284#discussioncomment-8760973). Let me provide a detailed explanation of the steps and commands required to run the main.py file in Docker:

1. Use _"docker ps"_ to list all running Docker containers and their IDs.

2. Execute _"docker exec -it <container_id> /bin/bash"_ to enter the file system of the Docker container. This allows access to files and the ability to run commands within the container.

3. In a new terminal, use _"docker cp /path_of_local_machineVC <container_id>:/path/to/new_created_folder_forVC"_ to copy the entire build visual circuit folder to the Docker container.

4. Return to the first terminal and run the main.py file.

(Note: It is not necessary to use the first terminal separately; you can also use the RoboticsAcademy web terminal, which is already inside the Docker file system.)

**Implementation of Autoparking using VisualCircuits:**

Referring to the [Autoparking documentation](https://jderobot.github.io/RoboticsAcademy/exercises/AutonomousCars/autoparking/), the Robot API section provides three laser data attributes: front, right, and back, as well as methods such as getPose, setW, and setV. Additional information about laser attributes can be found inside the laser attributes section. To gain a deeper understanding of these methods, try utilizing them within the RoboticsAcademy web IDE. To obtain topic names and types, navigate to the Autoparking exercise folder of RoboticsAcademy. With this information, we are ready to create logic using VisualCircuits.

In my demonstration, I used custom block for laser ROS1 Neotic, a Motor block, and a Code block. Since the Docker autoparking exercise runs on ROS1 Neotic, there was no predefined block for laser specific to ROS1 Neotic. The circuit is depicted below in the image.

![](https://github.com/JdeRobot/VisualCircuit/assets/62588059/80ec020a-a22f-4fd9-ad2e-12f272b14044)

**Videos/ Result:**

Using Visual Circuits: [youtube video](https://youtu.be/l1PBAalpbs0?si=p1XBM3pUMdVv_IS_)
Using RoboticsAcademy web IDE : [youtube video](https://youtu.be/4hV6KkkZvJ0?si=L1P4PwJHvZeyVjfe)

**Comment by toshan-luktuke** [comment]
Great to see the exercise working! @BkPankaj if you edit the video of the VIsualCircuit execution in the style of the [Visual Lander Video](https://www.youtube.com/watch?v=J1JhnNOKe1o) and send it to me we can even upload it on the JdeRobot channel.

**Reply by BkPankaj** [reply]
Thank you @toshan-luktuke ! I will send the video to you.

**Comment by jmplaza** [comment]
Congratulations @BkPankaj , it is a nice demo. 

RoboticsAcademy is now moving to ROS2-Humble (release 5.2.x), maybe you could replicate this with the ROS2-Humble sensor blocks and motor blocks in VisualCircuit. This way we could move forward also to ROS2-Humble in VisualCircuit. What do you think?

**Reply by BkPankaj** [reply]
Thank you, @jmplaza sir. It's a great idea. I will work on ROS2 Humble, and in between, this ROS1 Noetic video will also be useful for someone who needs a demo for ROS1 Noetic.
